File: JDK11/opentraceability/src/main/java/opentraceability/models/common/address.java
package opentraceability.models.common;

import opentraceability.utility.Country;

public class Address extends AnonymizedAddress {

    public String address1 = null;
    public String address2 = null;
    public String city = null;
    public String state = null;
    public String county = null;
    @Override
    public String zipCode = null;
    @Override
    public Country country = null;

    @Override
    public String toString() {
        StringBuilder pieces = new StringBuilder();
        addIfNotNull(pieces, address1);
        addIfNotNull(pieces, city);
        addIfNotNull(pieces, state);
        addIfNotNull(pieces, county);
        addIfNotNull(pieces, zipCode);
        addIfNotNull(pieces, country != null ? country.abbreviation : null);

        String addressStr = pieces.toString();
        return addressStr;
    }

    private void addIfNotNull(StringBuilder sb, String s){
        if (s != null && !s.isBlank()) {
            if (sb.length() > 0){ sb.append(", "); }
            sb.append(s);
        }
    }

    @Override
    public boolean equals(Object obj) {

        if (obj == null){
            return false;
        }

        if (!(obj instanceof Address)){
            return false;
        }

        return this.equals((Address)obj);
    }

    public boolean equals(Address other) {

        if (other == null)
        {
            return false;
        }

        return (this.address1 == other.address1
                && this.address2 == other.address2
                && this.city == other.city
                && this.county == other.county
                && this.state == other.state
                && this.zipCode == other.zipCode
                && this.country == other.country);
    }

    @Override
    public int hashCode() {
        return this.toString().hashCode();
    }
}

File: CSharp/OpenTraceability/Models/Common/Address.cs
namespace OpenTraceability.Models.Common
{
    public class Address : AnonymizedAddress
    {
        public string Address1 { get; set; }
        public string Address2 { get; set; }
        public string City { get; set; }
        public string State { get; set; }
        public string County { get; set; }

        public override string ToString()
        {
            List<string> pieces = new List<string>()
            {
                Address1,
                Address2,
                City,
                County,
                State,
                ZipCode,
                Country?.Abbreviation
            };
            string addressStr = string.Join(", ", pieces.Where(p => !string.IsNullOrWhiteSpace(p)));
            return addressStr;
        }

        public override bool Equals(object obj)
        {
            if (obj == null)
            {
                return false;
            }

            if (!(obj is Address))
            {
                return false;
            }

            return this.Equals(obj as Address);
        }

        public bool Equals(Address other)
        {
            if (other == null)
            {
                return false;
            }

            return (this.Address1 == other.Address1
                && this.Address2 == other.Address2
                && this.City == other.City
                && this.County == other.County
                && this.State == other.State
                && this.ZipCode == other.ZipCode
                && this.Country == other.Country);
        }

        public override int GetHashCode()
        {
            return this.ToString().GetHashCode();
        }
    }
}